nginx.conf

```diff
@@ -0,0 +1,28 @@
+upstream nodejs {
+    server web:4201;
+}
+
+server {
+    listen 80;
+
+    root /static;
+
+    location / {
+        try_files $uri sv/$uri @backend;
```

Comment:
I'm not familiar with nginx, so I'm wondering how this works with other locales than `sv`, for instance `fi`?

In the build both `dist/app/browser/` and `dist/app/server/` have subfolders for each locale from where Express would serve files based on the locale. If no locale is present in the URL, is the default locale (`sv`) served? `proxy-server.js` handles cases where the URL is missing locale and redirects to the default locale: https://github.com/slsfi/digital-edition-frontend-ng/blob/main/proxy-server.js#L34

Another related but slightly different matter, which I might have to take up with Identio, is this:

I'm also wondering if we could use nginx to switch to the correct locale of the app instead of doing it in `proxy-server.js`. Currently we have modified `server.ts` so the `app()` function is exported and can be used in `proxy-server.js` to switch locale. But this is not the standard way of deploying a multilingual app in Angular. Compare the Angular guide which has an nginx example of an app without SSR, but the principle should be the same I think even if you have SSR: https://angular.io/guide/i18n-common-deploy#nginx-example

In a standard multilingual SSR app each locale is built into a separate app located at `dist/app/browser/<locale>/` and `dist/app/server/<locale>/`. You would run the SSR app in a specific locale by running `dist/app/server/<locale>/main.js`.

So can nginx serve the app from, for instance, `dist/app/server/sv/main.js` if the user has requested the `sv` version, and from `dist/app/server/fi/main.js` if the user has requested the `fi` version? And would this increase performance compared to the current setup where the app is run from `proxy-server.js`?

Reply:
> Another related but slightly different matter, which I might have to take up with Identio, is this:
> 
> I'm also wondering if we could use nginx to switch to the correct locale of the app instead of doing it in proxy-server.js. Currently we have modified server.ts so the app() function is exported and can be used in proxy-server.js to switch locale. But this is not the standard way of deploying a multilingual app in Angular. Compare the Angular guide which has an nginx example of an app without SSR, but the principle should be the same I think even if you have SSR: https://angular.io/guide/i18n-common-deploy#nginx-example
> 
> In a standard multilingual SSR app each locale is built into a separate app located at dist/app/browser/<locale>/ and dist/app/server/<locale>/. You would run the SSR app in a specific locale by running dist/app/server/<locale>/main.js.
> 
> So can nginx serve the app from, for instance, dist/app/server/sv/main.js if the user has requested the sv version, and from dist/app/server/fi/main.js if the user has requested the fi version? And would this increase performance compared to the current setup where the app is run from proxy-server.js?

After reading up on SSR in cojunction with i18n in Angular it looks like we are stuck with using `proxy-server.js` for the time being. Angular doesn't yet provide a built-in solution for apps utilizing both of these features. The same kind of hacky workarounds have been required since Angular 9: https://github.com/angular/angular-cli/issues/25726

Reply:
The way the configuration is written, nginx tries first to find a matching file based on the URL directly, then tries again in the `sv` subfolder, and if the full URL is not found in either the root or the `sv` subfolder, it forwards the request to node and node handles the request.

For example, a request for the file https://topelius.sls.fi/sv/assets/images/home-page-banner-portrait.jpg would serve sv/assets/images/home-page-banner-portrait.jpg directly from the root, a request for the file https://topelius.sls.fi/assets/images/home-page-banner-portrait.jpg would serve that same file, but with a few milliseconds of delay as nginx would first try to find an `assets` folder in the root before retrying inside the `sv` folder.

In this way, nginx serves entirely static content directly without involving node, improving responsiveness for all static content, but requests for dynamic content are forwarded to node.

Reply:
Adding `fi/$uri` would mean a slightly longer delay yes, but I am not sure it would actually be measurable. Nginx is incredibly well optimized.

I wrote the configuration to try `sv` if there is no language code, since that is also the behavior used in the node.js code - if no language is specified, `sv` is chosen by default and the user is redirected.

If we knew for sure that the node.js code always uses a language code in the URL, we could even simply use `try_files $uri @backend` - have nginx attempt to serve the static content, and if it doesn't find anything let node handle things. Node could then redirect if there is no language code, and nginx could then serve static content directly.

Reply:
Let's leave the nginx config as you have written it, at least until we see what kind of performance boost this will give us.